### ISSUE TITLE
vmagent  v1.95.1 cannot discover target，The sleep parameter in groupWatchersCleaner cannot be adjusted.

### ISSUE THREAD

**Issue by will-jg** [VictoriaMetrics/VictoriaMetrics#5539]
### Describe the bug

When the role is set to `endpoints` without specifying a `namespace` or other conditions, it defaults to enabling watches for both pods and services. With a large volume of pod data in the cluster, the time taken to retrieve and parse pod data has reached 22 seconds. This exceeds the two observation periods of the `groupWatchersCleaner`, which are 14 seconds in total, leading to a situation where `awsTotal` is zero and causing the query for the entire `groupWatcher` to be canceled. As a result, services cannot be discovered. The question is whether the observation period of the `groupWatchersCleaner`, which is 7 seconds, can be exposed as a parameter to allow users to adjust it based on their situation, as slow responses from the Kubernetes API could lead to similar issues.

```golang
func groupWatchersCleaner(gws map[string]*groupWatcher) {
	for {
                // Hoping that the check cycle can be made configurable.
		time.Sleep(7 * time.Second)
		groupWatchersLock.Lock()
		for key, gw := range gws {
			gw.mu.Lock()
			// Calculate the number of apiWatcher instances subscribed to gw.
			awsTotal := 0
			for _, uw := range gw.m {
				awsTotal += len(uw.aws) + len(uw.awsPending)
			}

			if awsTotal == 0 {
				// There are no API watchers subscribed to gw.
				// Stop all the urlWatcher instances at gw and drop gw from gws in this case,
				// but do it only on the second iteration in order to reduce urlWatcher churn
				// during scrape config reloads.
				if gw.noAPIWatchers {
					gw.cancel()
					delete(gws, key)
				} else {
					gw.noAPIWatchers = true
				}
			} else {
				gw.noAPIWatchers = false
			}
			gw.mu.Unlock()
		}
		groupWatchersLock.Unlock()
	}
}
```

### To Reproduce

The scrape configuration is as follows, and the interface time consumption details are also commented in the code.
```yaml
global:
    scrape_interval: 30s
    external_labels:
        cluster: m6-prod
        pin: jdtech
        prometheus: brolly/prometheus-0
        prometheus_replica: prometheus-prometheus-0-0
scrape_configs:
    - job_name: brolly/tpaas-zk-service-monitor-0/0
      honor_labels: false
      kubernetes_sd_configs:
        - role: endpoints
          kubeconfig_file: /etc/vmagent.kubeconfig
      scrape_interval: 30s
      scrape_align_interval: 30s
      metrics_path: /metrics
      scheme: http
      relabel_configs:
        - action: keep
          source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_app
          regex: zookeeper-cluster
        - action: keep
          source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_component
          regex: zk
        - action: keep
          source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_deploy_by
          regex: paas
        - action: keep
          source_labels:
            - __meta_kubernetes_endpoint_port_name
          regex: metrics
        - source_labels:
            - __meta_kubernetes_endpoint_address_target_kind
            - __meta_kubernetes_endpoint_address_target_name
          separator: ;
          regex: Node;(.*)
          replacement: ${1}
          target_label: node
        - source_labels:
            - __meta_kubernetes_endpoint_address_target_kind
            - __meta_kubernetes_endpoint_address_target_name
          separator: ;
          regex: Pod;(.*)
          replacement: ${1}
          target_label: pod
        - source_labels:
            - __meta_kubernetes_namespace
          target_label: namespace
        - source_labels:
            - __meta_kubernetes_service_name
          target_label: service
        - source_labels:
            - __meta_kubernetes_pod_name
          target_label: pod
        - source_labels:
            - __meta_kubernetes_pod_container_name
          target_label: container
        - source_labels:
            - __meta_kubernetes_service_name
          target_label: job
          replacement: ${1}
        - target_label: endpoint
          replacement: metrics
        - source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_name
          target_label: app_name
          action: replace
        - source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_instance
          target_label: app_instance
          action: replace
        - source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_app
          target_label: app
          action: replace
        - source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_component
          target_label: app_component
          action: replace
        - source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_managed_by
          target_label: app_managed_by
          action: replace
        - source_labels:
            - __meta_kubernetes_service_label_app_kubernetes_io_deploy_by
          target_label: app_deploy_by
          action: replace
        - regex: __meta_kubernetes_pod_label_(.+)
          action: labelmap
        - source_labels:
            - app_component
          target_label: integration
          replacement: tpaas-zk
          action: replace
        - source_labels:
            - __meta_kubernetes_pod_label_account_jdos_io_pin
          target_label: pin
          action: replace
```
```golang
func (uw *urlWatcher) reloadObjects() string {
	if uw.resourceVersion != "" {
		// Fast path - there is no need in reloading the objects.
		return uw.resourceVersion
	}

	gw := uw.gw
	startTime := time.Now()
	apiURL := uw.apiURL

	delimiter := getQueryArgsDelimiter(apiURL)
	requestURL := apiURL + delimiter + "resourceVersion=0&resourceVersionMatch=NotOlderThan"
// Requesting to start.
	resp, err := gw.doRequest(gw.ctx, requestURL)
// Requesting to end .  Time-consuming 8s.
	if err != nil {
		if !errors.Is(err, context.Canceled) {
			logger.Errorf("cannot perform request to %q: %s", requestURL, err)
		}
		logger.Errorf("cannot perform request to2 %q: %s", requestURL, err)
		return ""
	}
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		_ = resp.Body.Close()
		logger.Errorf("unexpected status code for request to %q: %d; want %d; response: %q", requestURL, resp.StatusCode, http.StatusOK, body)
		return ""
	}

// Parse to end . 
	objectsByKey, metadata, err := uw.parseObjectList(resp.Body)
// Parse to end .  Time-consuming 12s.  
// A total of 20 seconds. Exceeded two check cycles 14 seconds
// uw.aws is still empty at this time.
	
// ....

	uw.removeScrapeWorksLocked(objectsRemoved)
	uw.updateScrapeWorksLocked(objectsUpdated, uw.aws)
	uw.updateScrapeWorksLocked(objectsAdded, uw.aws)
	uw.needRecreateScrapeWorks = false
	if len(objectsRemoved) > 0 || len(objectsUpdated) > 0 || len(objectsAdded) > 0 {
		uw.maybeUpdateDependedScrapeWorksLocked()
	}
	gw.mu.Unlock()

	uw.objectsUpdated.Add(len(objectsUpdated))
	uw.objectsRemoved.Add(len(objectsRemoved))
	uw.objectsAdded.Add(len(objectsAdded))
	uw.objectsCount.Add(len(objectsAdded) - len(objectsRemoved))
	uw.resourceVersion = metadata.ResourceVersion

	logger.Infof("reloaded %d objects from %q in %.3fs; updated=%d, removed=%d, added=%d, resourceVersion=%q",
		len(objectsByKey), requestURL, time.Since(startTime).Seconds(), len(objectsUpdated), len(objectsRemoved), len(objectsAdded), uw.resourceVersion)
	return uw.resourceVersion
}
```

### Version

vmagent-20231116-193721-tags-v1.95.1-0-g354563393

### Logs

After increasing the sleep period in groupWatchersCleaner to 15 seconds, service discovery operates normally, outputting the following log, taking 19 seconds.

![1703669118031](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/121209484/365caa9d-cfe2-4d72-8162-90262e9af2ac)

**Comment by hagen1778** [comment]
@zekker6 or @Haleygo could you please assist with this question? Maybe, instead of exposing the extra config we can try checking if there are pending/unfinished requests to k8s API for this wathcer?

**Comment by valyala** [comment]
@will-jg , thanks for filing the detailed issue!

It should be fixed by the commit 4e3242b02da0a814274bf08a4f5a49f86a785cbe . This commit has been included in [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html) and [v1.93.11 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.11). Closing the issue as fixed.

@will-jg , please re-open the issue if it is still reproduced in the [latest vmagent release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/latest).
